#Pipeline output script
#Code partially based on the code from https://github.com/fiererlab/dada2_fiererlab with modification  by George Cox
#Takes outputs from pipeline and prepares them for downstream analyses such as phylogeny, community analysis, and pipeline testing.
#Follows on from the modular DADA2 pipeline (DADA2_pipeline_DATE.R).
#Doesn't need to be run on a VM as it only takes about a minute to run with all systems going. 
#Can be run immediately after a pipeline run overnight as part of the modular DADA2 pipeline (code in the DADA2 pipeline).


#Variables and Parameters block (CHECK DATES AND FILEPATHS!)
#__________________________________________________________________________________________

Date <- "1_8_23" #Date when run started
Inpath <- "~/outputs/2_8_23" #Where the input files are stored, often the output of a pipeline run
setwd(Inpath) #sets working directory so you can be lazy and not type in file paths for the inputs.

contig_format <- TRUE #Turns on or off the formatting of the contig pipeline files.
seqtab_contig_input <- readRDS("seqtab_nochim_contig_1_8_23.rds") #The seqtab generated by the contig pipeline (or any normal dada2 pipeline), Reccomended to use the no chimera seqtab. 
taxa_contig_input <- readRDS("taxa_contig_1_8_23.rds") #Reading in the taxonomy table from the contig pipeline (or any normal dada2 pipeline).

forward_only_format <- TRUE #Turns on/off the formatting of the forward only pipeline files.
seqtab_forward_only_input <- readRDS("seqtab_forward_only_1_8_23.rds") #reads in the seqtab from the forward only pipeline. Reccomended to use the no chimera seqtab. 
taxa_forward_only_input <- readRDS("taxa_forward_only_1_8_23.rds") #reads in the forward only taxonomy table. 

N_mash_formatting <- TRUE #Turns on/off the formatting of the N mash pipeline files.
seqtab_N_mash_input <- readRDS("seqtab_N_mash_1_8_23.rds") #seqtab for N mash pipeline, use no chim seqtab.
taxa_N_mash_input <- readRDS("taxa_N_mash_1_8_23.rds") #reads in the N mash taxonomy table. 


print(Date)
print(Inpath)
print(contig_format)
print(forward_only_format)
print(N_mash_formatting)


#Packages block
#________________________________________________________________________________________

#install.packages("dplyr")
#install.packages("tidyr")
#install.packages("stringr")
library(dplyr)
library(tidyr)
library(stringr)

#Functions block
#________________________________________________________________________________________

writeRepSetFasta<-function(data, filename){
  fastaLines = c()
  for (rowNum in 1:nrow(data)){
    fastaLines = c(fastaLines, as.character(paste(">", data[rowNum,"name"], sep = "")))
    fastaLines = c(fastaLines,as.character(data[rowNum,"seq"]))
  }
  fileConn<-file(filename)
  writeLines(fastaLines, fileConn)
  close(fileConn)
}


#To run on VM
#________________________________________________________________________________________


#Code block
#________________________________________________________________________________________
#Reformat files into more useful things for later parts
dir.create("reform_outputs")


if(contig_format == TRUE){ #Code from Fierer lab github.
  seqtab_t_contig <- as.data.frame(t(seqtab_contig_input)) #flips seqtab to be a bit more readable
  rep_set_ASVs_contig <- as.data.frame(rownames(seqtab_t_contig)) #Pulls out the names of the ASVs (currently their sequence which we want to change)
  rep_set_ASVs_contig <- mutate(rep_set_ASVs_contig, ASV_ID = 1:n()) #Names the asvs, with more abundant ones have lower numbers
  rep_set_ASVs_contig$ASV_ID <- sub("^", "ASV_", rep_set_ASVs_contig$ASV_ID) #renames the asvs to be the ASV_ID, the sub function uses regex to find and replace the name using the ^ wildcard.
  rep_set_ASVs_contig$ASV <- rep_set_ASVs_contig$`rownames(seqtab_t_contig)` #Adds the site/sample names to the ASV rep set
  rep_set_ASVs_contig$`rownames(seqtab_t_contig)` <- NULL #Cleans up the ASV repset by getting rid of the original rownames. 
  
  rownames(seqtab_t_contig) <- rep_set_ASVs_contig$ASV_ID #adds the new row names to the ASV table
  saveRDS(seqtab_t_contig, file = "reform_outputs/seqtab_contig_with_ASV_IDs_DATE.rds")
  
  taxonomy_contig <- as.data.frame(taxa_contig_input)
  taxonomy_contig$ASV <- as.factor(rownames(taxonomy_contig))
  taxonomy_contig <- merge(rep_set_ASVs_contig, taxonomy_contig, by = "ASV")
  rownames(taxonomy_contig) <- taxonomy_contig$ASV_ID
  taxonomy_contig$Species <- gsub("\\.", "", taxonomy_contig$Species) #removes all . (full stops) in the code so that downstream programs dont have issues with the fasta file. 
  taxonomy_contig$Genus <- gsub("\\.", "", taxonomy_contig$Genus) #Same as above but for Genus
  
  
  taxonomy_for_fasta_contig <- taxonomy_contig %>%
    unite("TaxString", c("Domain", "Supergroup", "Division", "Kingdom", "Phylum", "Order","Family", "Genus", "Species"), 
          sep = ";", remove = FALSE) %>%
    unite("name", c("ASV_ID", "TaxString"), 
          sep = "__", remove = TRUE) %>%
    select(ASV, name) %>%
    rename(seq = ASV)
  writeRepSetFasta(taxonomy_for_fasta_contig,"reform_outputs/repset_contig_taxonomy_DATE.fasta")
  
  #Mash tables and write them to disk
  seqtab_taxa_contig <- merge(seqtab_t_contig, taxonomy_contig, by = 0)
  rownames <- (seqtab_taxa_contig$Row.names)
  rownames(seqtab_taxa_contig) <- rownames
  seqtab_taxa_contig$Row.names <- NULL
  saveRDS(seqtab_taxa_contig, "reform_outputs/seqtab_taxa_contig_DATE.rds")
  
  
  #filter out bacteria, archaea, fungi, plants, and animals
  pre_filter <- nrow(taxonomy_contig)
  filtered_taxonomy_contig <- taxonomy_contig[ grep("Bacteria", taxonomy_contig$Domain, invert = TRUE),]#Removes all bacteria
  bacteria_loss <- nrow(filtered_taxonomy_contig)
  filtered_taxonomy_contig <- filtered_taxonomy_contig[ grep("Archaea", filtered_taxonomy_contig$Domain, invert = TRUE),] #Removes all archaea
  archaea_loss <- nrow(filtered_taxonomy_contig)
  filtered_taxonomy_contig <- filtered_taxonomy_contig[ grep("Fungi", filtered_taxonomy_contig$Kingdom, invert = TRUE),] #Removes fungi
  fungi_loss <- nrow(filtered_taxonomy_contig)
  filtered_taxonomy_contig <- filtered_taxonomy_contig[ grep("Metazoa", filtered_taxonomy_contig$Kingdom, invert = TRUE),] #Removes animals
  animal_loss <- nrow(filtered_taxonomy_contig)
  filtered_taxonomy_contig <- filtered_taxonomy_contig[ grep("Embryophyceae", filtered_taxonomy_contig$Phylum, invert = TRUE),] #removes plants
  plant_loss <- nrow(filtered_taxonomy_contig)
  filter_stage <- c("Pre filter", "Bacteria filter", "Archaea filter", "Fungi filter", "Animal filter", "Plant filter")
  taxonomy_df_size <- c(pre_filter, bacteria_loss, archaea_loss, fungi_loss, animal_loss, plant_loss)
  filter_impact_contig <- data.frame(filter_stage, taxonomy_df_size)
  saveRDS(filter_impact_contig, "reform_outputs/filter_impact_contig_DATE.rds")
  
  filtered_taxonomy_for_fasta_contig <- filtered_taxonomy_contig %>% #Writes a fasta file without the non target groups. 
    unite("TaxString", c("Domain", "Supergroup", "Division", "Kingdom", "Phylum", "Order","Family", "Genus", "Species"), 
          sep = ";", remove = FALSE) %>%
    unite("name", c("ASV_ID", "TaxString"), 
          sep = "__", remove = TRUE) %>%
    select(ASV, name) %>%
    rename(seq = ASV)
  writeRepSetFasta(filtered_taxonomy_for_fasta_contig,"reform_outputs/repset_filtered_contig_taxonomy_DATE.fasta")
  
  
  
  #Merging seqtab and filtered taxonomy
  seqtab_taxa_filtered_contig <- merge(seqtab_t_contig, filtered_taxonomy_contig, by = 0)
  rownames <- (seqtab_taxa_filtered_contig$Row.names)
  rownames(seqtab_taxa_filtered_contig) <- rownames
  seqtab_taxa_filtered_contig$Row.names <- NULL
  saveRDS(seqtab_taxa_filtered_contig, "reform_outputs/seqtab_taxa_filtered_contig_DATE.rds")
 
  print("Contig Formatting Finished") 
}



#Make asv fasta for phylogeny



if(forward_only_format == TRUE){
  seqtab_t_forward_only <- as.data.frame(t(seqtab_forward_only_input)) #flips seqtab to be a bit more readable
  rep_set_ASVs_forward_only <- as.data.frame(rownames(seqtab_t_forward_only)) #Pulls out the names of the ASVs (currently their sequence which we want to change)
  rep_set_ASVs_forward_only <- mutate(rep_set_ASVs_forward_only, ASV_ID = 1:n()) #Names the asvs, with more abundant ones have lower numbers
  rep_set_ASVs_forward_only$ASV_ID <- sub("^", "ASV_", rep_set_ASVs_forward_only$ASV_ID) #renames the asvs to be the ASV_ID, the sub function uses regex to find and replace the name using the ^ wildcard.
  rep_set_ASVs_forward_only$ASV <- rep_set_ASVs_forward_only$`rownames(seqtab_t_forward_only)` #Adds the site/sample names to the ASV rep set
  rep_set_ASVs_forward_only$`rownames(seqtab_t_forward_only)` <- NULL #Cleans up the ASV repset by getting rid of the original rownames. 
  
  rownames(seqtab_t_forward_only) <- rep_set_ASVs_forward_only$ASV_ID #adds the new row names to the ASV table
  saveRDS(seqtab_t_forward_only, file = "reform_outputs/seqtab_forward_only_with_ASV_IDs_DATE.rds")
  
  taxonomy_forward_only <- as.data.frame(taxa_forward_only_input)
  taxonomy_forward_only$ASV <- as.factor(rownames(taxonomy_forward_only))
  taxonomy_forward_only <- merge(rep_set_ASVs_forward_only, taxonomy_forward_only, by = "ASV")
  rownames(taxonomy_forward_only) <- taxonomy_forward_only$ASV_ID
  taxonomy_forward_only$Species <- gsub("\\.", "", taxonomy_forward_only$Species) #removes all . (full stops) in the code so that downstream programs dont have issues with the fasta file. 
  taxonomy_forward_only$Genus <- gsub("\\.", "", taxonomy_forward_only$Genus) #Same as above but for Genus
  
  
  taxonomy_for_fasta_forward_only <- taxonomy_forward_only %>%
    unite("TaxString", c("Domain", "Supergroup", "Division", "Kingdom", "Phylum", "Order","Family", "Genus", "Species"), 
          sep = ";", remove = FALSE) %>%
    unite("name", c("ASV_ID", "TaxString"), 
          sep = "__", remove = TRUE) %>%
    select(ASV, name) %>%
    rename(seq = ASV)
  writeRepSetFasta(taxonomy_for_fasta_forward_only,"reform_outputs/repset_forward_only_taxonomy_DATE.fasta")
  
  #Mash tables and write them to disk
  seqtab_taxa_forward_only <- merge(seqtab_t_forward_only, taxonomy_forward_only, by = 0)
  rownames <- (seqtab_taxa_forward_only$Row.names)
  rownames(seqtab_taxa_forward_only) <- rownames
  seqtab_taxa_forward_only$Row.names <- NULL
  saveRDS(seqtab_taxa_forward_only, "reform_outputs/seqtab_taxa_forward_only_DATE.rds")
  
  
  #filter out bacteria, archaea, fungi, plants, and animals
  pre_filter <- nrow(taxonomy_forward_only)
  filtered_taxonomy_forward_only <- taxonomy_forward_only[ grep("Bacteria", taxonomy_forward_only$Domain, invert = TRUE),]#Removes all bacteria
  bacteria_loss <- nrow(filtered_taxonomy_forward_only)
  filtered_taxonomy_forward_only <- filtered_taxonomy_forward_only[ grep("Archaea", filtered_taxonomy_forward_only$Domain, invert = TRUE),] #Removes all archaea
  archaea_loss <- nrow(filtered_taxonomy_forward_only)
  filtered_taxonomy_forward_only <- filtered_taxonomy_forward_only[ grep("Fungi", filtered_taxonomy_forward_only$Kingdom, invert = TRUE),] #Removes fungi
  fungi_loss <- nrow(filtered_taxonomy_forward_only)
  filtered_taxonomy_forward_only <- filtered_taxonomy_forward_only[ grep("Metazoa", filtered_taxonomy_forward_only$Kingdom, invert = TRUE),] #Removes animals
  animal_loss <- nrow(filtered_taxonomy_forward_only)
  filtered_taxonomy_forward_only <- filtered_taxonomy_forward_only[ grep("Embryophyceae", filtered_taxonomy_forward_only$Phylum, invert = TRUE),] #removes plants
  plant_loss <- nrow(filtered_taxonomy_forward_only)
  filter_stage <- c("Pre filter", "Bacteria filter", "Archaea filter", "Fungi filter", "Animal filter", "Plant filter")
  taxonomy_df_size <- c(pre_filter, bacteria_loss, archaea_loss, fungi_loss, animal_loss, plant_loss)
  filter_impact_forward_only <- data.frame(filter_stage, taxonomy_df_size)
  saveRDS(filter_impact_forward_only, "reform_outputs/filter_impact_forward_only_DATE.rds")
  
  filtered_taxonomy_for_fasta_forward_only <- filtered_taxonomy_forward_only %>% #Writes a fasta file without the non target groups. 
    unite("TaxString", c("Domain", "Supergroup", "Division", "Kingdom", "Phylum", "Order","Family", "Genus", "Species"), 
          sep = ";", remove = FALSE) %>%
    unite("name", c("ASV_ID", "TaxString"), 
          sep = "__", remove = TRUE) %>%
    select(ASV, name) %>%
    rename(seq = ASV)
  writeRepSetFasta(filtered_taxonomy_for_fasta_forward_only,"reform_outputs/repset_filtered_forward_only_taxonomy_DATE.fasta")
  
  #Merging seqtab and filtered taxonomy
  seqtab_taxa_filtered_forward_only <- merge(seqtab_t_forward_only, filtered_taxonomy_forward_only, by = 0)
  rownames <- (seqtab_taxa_filtered_forward_only$Row.names)
  rownames(seqtab_taxa_filtered_forward_only) <- rownames
  seqtab_taxa_filtered_forward_only$Row.names <- NULL
  saveRDS(seqtab_taxa_filtered_forward_only, "reform_outputs/seqtab_taxa_filtered_forward_only_DATE.rds")
  
  print("forward_only Formatting Finished") 
}


if(N_mash_formatting == TRUE){
  seqtab_t_N_mash <- as.data.frame(t(seqtab_N_mash_input)) #flips seqtab to be a bit more readable
  rep_set_ASVs_N_mash <- as.data.frame(rownames(seqtab_t_N_mash)) #Pulls out the names of the ASVs (currently their sequence which we want to change)
  rep_set_ASVs_N_mash <- mutate(rep_set_ASVs_N_mash, ASV_ID = 1:n()) #Names the asvs, with more abundant ones have lower numbers
  rep_set_ASVs_N_mash$ASV_ID <- sub("^", "ASV_", rep_set_ASVs_N_mash$ASV_ID) #renames the asvs to be the ASV_ID, the sub function uses regex to find and replace the name using the ^ wildcard.
  rep_set_ASVs_N_mash$ASV <- rep_set_ASVs_N_mash$`rownames(seqtab_t_N_mash)` #Adds the site/sample names to the ASV rep set
  rep_set_ASVs_N_mash$`rownames(seqtab_t_N_mash)` <- NULL #Cleans up the ASV repset by getting rid of the original rownames. 
  
  rownames(seqtab_t_N_mash) <- rep_set_ASVs_N_mash$ASV_ID #adds the new row names to the ASV table
  saveRDS(seqtab_t_N_mash, file = "reform_outputs/seqtab_N_mash_with_ASV_IDs_DATE.rds")
  
  taxonomy_N_mash <- as.data.frame(taxa_N_mash_input)
  taxonomy_N_mash$ASV <- as.factor(rownames(taxonomy_N_mash))
  taxonomy_N_mash <- merge(rep_set_ASVs_N_mash, taxonomy_N_mash, by = "ASV")
  rownames(taxonomy_N_mash) <- taxonomy_N_mash$ASV_ID
  taxonomy_N_mash$Species <- gsub("\\.", "", taxonomy_N_mash$Species) #removes all . (full stops) in the code so that downstream programs dont have issues with the fasta file. 
  taxonomy_N_mash$Genus <- gsub("\\.", "", taxonomy_N_mash$Genus) #Same as above but for Genus
  
  
  taxonomy_for_fasta_N_mash <- taxonomy_N_mash %>%
    unite("TaxString", c("Domain", "Supergroup", "Division", "Kingdom", "Phylum", "Order","Family", "Genus", "Species"), 
          sep = ";", remove = FALSE) %>%
    unite("name", c("ASV_ID", "TaxString"), 
          sep = "__", remove = TRUE) %>%
    select(ASV, name) %>%
    rename(seq = ASV)
  writeRepSetFasta(taxonomy_for_fasta_N_mash,"reform_outputs/repset_N_mash_taxonomy_DATE.fasta")
  
  #Mash tables and write them to disk
  seqtab_taxa_N_mash <- merge(seqtab_t_N_mash, taxonomy_N_mash, by = 0)
  rownames <- (seqtab_taxa_N_mash$Row.names)
  rownames(seqtab_taxa_N_mash) <- rownames
  seqtab_taxa_N_mash$Row.names <- NULL
  saveRDS(seqtab_taxa_N_mash, "reform_outputs/seqtab_taxa_N_mash_DATE.rds")
  
  
  #filter out bacteria, archaea, fungi, plants, and animals
  pre_filter <- nrow(taxonomy_N_mash)
  filtered_taxonomy_N_mash <- taxonomy_N_mash[ grep("Bacteria", taxonomy_N_mash$Domain, invert = TRUE),]#Removes all bacteria
  bacteria_loss <- nrow(filtered_taxonomy_N_mash)
  filtered_taxonomy_N_mash <- filtered_taxonomy_N_mash[ grep("Archaea", filtered_taxonomy_N_mash$Domain, invert = TRUE),] #Removes all archaea
  archaea_loss <- nrow(filtered_taxonomy_N_mash)
  filtered_taxonomy_N_mash <- filtered_taxonomy_N_mash[ grep("Fungi", filtered_taxonomy_N_mash$Kingdom, invert = TRUE),] #Removes fungi
  fungi_loss <- nrow(filtered_taxonomy_N_mash)
  filtered_taxonomy_N_mash <- filtered_taxonomy_N_mash[ grep("Metazoa", filtered_taxonomy_N_mash$Kingdom, invert = TRUE),] #Removes animals
  animal_loss <- nrow(filtered_taxonomy_N_mash)
  filtered_taxonomy_N_mash <- filtered_taxonomy_N_mash[ grep("Embryophyceae", filtered_taxonomy_N_mash$Phylum, invert = TRUE),] #removes plants
  plant_loss <- nrow(filtered_taxonomy_N_mash)
  filter_stage <- c("Pre filter", "Bacteria filter", "Archaea filter", "Fungi filter", "Animal filter", "Plant filter")
  taxonomy_df_size <- c(pre_filter, bacteria_loss, archaea_loss, fungi_loss, animal_loss, plant_loss)
  filter_impact_N_mash <- data.frame(filter_stage, taxonomy_df_size)
  saveRDS(filter_impact_N_mash, "reform_outputs/filter_impact_N_mash_DATE.rds")
  
  filtered_taxonomy_for_fasta_N_mash <- filtered_taxonomy_N_mash %>% #Writes a fasta file without the non target groups. 
    unite("TaxString", c("Domain", "Supergroup", "Division", "Kingdom", "Phylum", "Order","Family", "Genus", "Species"), 
          sep = ";", remove = FALSE) %>%
    unite("name", c("ASV_ID", "TaxString"), 
          sep = "__", remove = TRUE) %>%
    select(ASV, name) %>%
    rename(seq = ASV)
  writeRepSetFasta(filtered_taxonomy_for_fasta_N_mash,"reform_outputs/repset_filtered_N_mash_taxonomy_DATE.fasta")
  
  #Merging seqtab and filtered taxonomy
  seqtab_taxa_filtered_N_mash <- merge(seqtab_t_N_mash, filtered_taxonomy_N_mash, by = 0)
  rownames <- (seqtab_taxa_filtered_N_mash$Row.names)
  rownames(seqtab_taxa_filtered_N_mash) <- rownames
  seqtab_taxa_filtered_N_mash$Row.names <- NULL
  saveRDS(seqtab_taxa_filtered_N_mash, "reform_outputs/seqtab_taxa_filtered_N_mash_DATE.rds")
  
  print("N_mash Formatting Finished") 
}


#Output clean up block
#_______________________________________________________________________________


setwd("reform_outputs")
file.copy(from = '~/pipelines/Pipeline_formatting_DATE.R', to = 'Pipeline_formatting_DATE.R')
file.rename(list.files(pattern = "DATE"), str_replace(list.files(pattern = "DATE"),pattern = "DATE", Date))


